File: 0744-find-smallest-letter-greater-than-target/0744-find-smallest-letter-greater-than-target.rb
# @param {Character[]} letters
# @param {Character} target
# @return {Character}
def next_greatest_letter(letters, target)
    low = 0
    high = letters.size-1
    while(low <= high)
        mid = (low + high) / 2
        if(letters[mid] > target)
          high = mid - 1
        else
          low = mid + 1
        end
    end
    return letters[low % letters.size]
end

File: 0744-find-smallest-letter-greater-than-target/0744-find-smallest-letter-greater-than-target.cpp
int init=[] {ios_base::sync_with_stdio(false);cin.tie(0);return 0;}();
class Solution {
public:
    char nextGreatestLetter(vector<char>& letters, char target) {
        if (target >= letters.back()) return letters[0] ;
        int low = 0, high = letters.size();
    
        while( low < high) {
            int mid = low + (high-low) / 2;
            if (letters[mid] <= target) {
                low = mid + 1; 
            } else {
                high = mid;
            }
        }

        return letters[low];        
    }
};